### PR TITLE
Added CALL and RET operators for function-like control flow.

### PR DESCRIPTION
**Pull request by pianoman373**
This adds the CALL and RET keywords to TIS-3D. They work the same as in normal assembly language (I think), and didn't take all that many changes to implement. 

One notable change is the addition of a new variable to MachineState which stores the return address (reta) that RET should jump to.

### CODE REVIEW

**Comment by fnuecke** [comment]
Thanks! But... a couple of things.

Given the length of TIS-3D programs, I'm not fully convinced this is worth the added complexity as is.

From my experience typically, `CALL` and `RET` work by pushing the old `PC` onto a stack in RAM, the head of which is pointed to by a stack pointer register. Which isn't a thing in TIS-3D. Working around that by adding a register just to hold the PC to return to is not great, because that way multiple calls will break things (beginning with the second `RET` there will be undefined behavior). And changing that to a stack gives too much storage capabilities to the execution module IMHO.

There might still be a way to salvage this and fix both of the above. And that would be to allow using the stack module as the actual call stack. There's two ways I can think of:
- have a register defining the side of the stack to use, and an instruction setting that side (e.g. `STACK LEFT`),
- pass the side to use as a stack along in the call and return instructions (i.e. `CALL label LEFT`/`CALL LEFT label` and `RET LEFT`).

The first one is more verbose, as in it's eating at least one line, second would need more internal reworkings for the parser I think, and is more error prone. So I'm actually favoring the first approach. This would also make `PUSH` and `POP` keywords an option.

This re-uses stuff that's already in the mod, which is great, and adds even more potential for creativity, because, well, you don't have to use a _stack_ module on that side, after all ;)

So, TL;DR I'd say: change `reta` to `Port stack`, defining the side the stack is on, add `STACK` instruction to change it, write to/read from that side when calling/returning.

Possibly: add push/pop instructions, a la `PUSH src`, where `src` can be a side or an immediate value. So push/pop would basically be side-bound moves. Which is why I'm not super sure it'd be worth having them.

Oh, also, you missed a println in the return instruction.

**Comment by pianoman373** [comment]
Ohhhhhhh.......... I totally forgot about nested functions. Well the first approach sounds good, but like you said it may be too much complexity for such a small feature. For instance, what would happen if CALL or any of the others were used but STACK was never used. It would have to do nothing I guess and that would be a lot more confusing.

Another approach worth mentioning would be the ability to jump absolutely from a register. Just like how JRO works but with 0 being the beginning of the program. Then it could use $ as a read-only register that has the program counter variable. Doing something like this would work the same as a CALL as long as a stack is connected on the left. Let's say JAB is jump absolute.

MOV $ ACC
ADD $ 4
MOV ACC LEFT
JMP FUNCTION
#where program resumes

FUNCTION:
#do stuff
JAB LEFT

This is a much more complicated way to do it, but the added features would be more general purpose. I definitely see now though how complicated it would be and it might not be worth adding.

**Comment by pianoman373** [comment]
Yeah it looks like at this point, any alternatives would be a new commit entirely, and would be a much bigger project. Closing PR.
